File: BOJ/sort/24061.java
import java.io.*;
import java.util.*;

public class Main {

    static boolean finished = false;
    static int sortedCnt = 0;
    static int k;

	public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        String[] inputs = br.readLine().split(" ");
        int n = Integer.parseInt(inputs[0]);
        k = Integer.parseInt(inputs[1]);
        int[] arr = new int[n];

        inputs = br.readLine().split(" ");
        for (int i=0; i<inputs.length; i++) {
            int v = Integer.parseInt(inputs[i]);
            arr[i] = v;
        }

        mergeSort(arr, 0, arr.length-1);

        if (!finished) {
            System.out.println(-1);
        }
    }

    static void mergeSort(int[] arr, int p, int r) {
        if (p >= r || finished) return;
        int q = (p + r) / 2;
        // System.out.printf("p:%d q:%d r:%d\n", p, q, r);
        mergeSort(arr, p, q);
        mergeSort(arr, q+1, r);
        merge(arr, p, q, r);
    }

    static void merge(int[] arr, int p, int q, int r) {
        if (finished) return;
        int[] tmp = new int[r - p + 1];
        
        int i = p, j = q+1, t = 0;
        // 큰 순으로 각 포인터에서 처리
        while (i <= q && j <= r) {
            if (arr[i] <= arr[j]) tmp[t++] = arr[i++];
            else tmp[t++] = arr[j++];
        }

        // 남은 것 처리
        while (i <= q) tmp[t++] = arr[i++];
        while (j <= r) tmp[t++] = arr[j++];

        for (i=0; i<t; i++) {
            arr[p + i] = tmp[i];

            if (!finished && ++sortedCnt == k) {
                finished = true;
                for (int a : arr) {
                    System.out.printf("%d ", a);
                }
                System.out.println();
                break;
            }
        }
    }
}

File: BOJ/sort/24061.cpp
#include <bits/stdc++.h>

using namespace std;

#define FASTIO cin.tie(0); cout.tie(0); ios_base::sync_with_stdio(0);

typedef long long ll;
typedef unsigned long long ull;
typedef pair<int, int> ii;
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef vector<ii> vii;
const int INF = 987654321;

bool finished = false;
int sortedCnt = 0;
int k;

void merge(vector<int>& arr, int p, int q, int r) {
    if (finished) return;

    vector<int> tmp;
    tmp.reserve(r - p + 1);

    int i = p, j = q + 1;

    while (i <= q && j <= r) {
        if (arr[i] <= arr[j]) tmp.push_back(arr[i++]);
        else tmp.push_back(arr[j++]);
    }

    while (i <= q) tmp.push_back(arr[i++]);
    while (j <= r) tmp.push_back(arr[j++]);

    for (int t = 0; t < (int)tmp.size(); t++) {
        arr[p + t] = tmp[t];

        if (!finished && ++sortedCnt == k) {
            finished = true;
            for (int v : arr) cout << v << ' ';
            cout << '\n';
            return;
        }
    }
}

void mergeSort(vector<int>& arr, int p, int r) {
    if (p >= r || finished) return;

    int q = (p + r) / 2;
    mergeSort(arr, p, q);
    mergeSort(arr, q + 1, r);
    merge(arr, p, q, r);
}

int main() {
    FASTIO;

    int n; cin >> n >> k;

    vector<int> arr(n);
    for (int i = 0; i < n; i++) cin >> arr[i];

    mergeSort(arr, 0, n - 1);

    if (!finished) cout << -1 << '\n';

    return 0;
}